data/lua/wml/harm_unit.lua

```diff
@@ -168,11 +194,15 @@ function wml_actions.harm_unit(cfg)
 			if experience ~= false and harmer and harmer.valid
 				and wesnoth.sides.is_enemy( unit_to_harm.side, harmer.side )
 			then
-				if kill ~= false and unit_to_harm.hitpoints <= 0 then
+				if kill ~= false and unit_to_harm.hitpoints <= 0 and xp_mode.kill then
```

Comment:
this part is wrong then, xp_mode.kill needs to be wrapped in new if

Reply:
The current behavior makes sense to me, that change doesn't.

* Get 2 XP for attacking a level 2 unit, regardless of whether the attack kills, vs
* Get 2 XP for attacking a level 2 unit, but get 0 XP instead if the attack kills.

Reply:
If there is no experience=attack/fight, then fight xp should not be given.

Reply:
This is with `experience=attack`. If Bob has 5 hitpoints, and Alice does 4 damage, then Alice gets XP; the question is whether Alice should get 0 XP if she does more damage.

Reply:
From player perspective it would make sense to want xp, but non-harm_unit fight would not give xp.

Reply:
I'm confused by the edit - I'm assuming we're discussing the situation with `wesnoth.game_config.kill_experience` at its default value. If someone reduced `kill_experience` to zero, why wouldn't they also reduce the attack experience to zero at the same time, thus avoiding an situation where doing more damage would result in less XP?

Reply:
A reason for that can be lore, if we have mind eater that earns xp only by eating living minds then killing shouldn't give it any xp, but attaching should.